### PR TITLE
add Capture-ID message consolidation to call flow

### PR DESCRIPTION
**Pull request by danieldonoghue**
This change allows the UI to group messages in the Call Flow tab such that all
rows with the same fingerprint but from different Capture IDs are presented as
one, with a dropdown to access the individual messages.

In multi-sensor deployments, multiple capture agents frequently see the same SIP
message. Without consolidation this creates duplicate rows in the call flow,
which can obscure the actual sequence of events and make the diagram harder to
read. This feature collapses those duplicates into a single row with a +N badge
and an expand control, so the flow chart stays clean without losing access to
the per-agent detail.

The fingerprint is computed on the fly in the UI from each message's source IP,
source port, destination IP, destination port, and payload so no ingest-side
changes or server configuration are required. If a pre-computed fingerprint
field is already present on a message row it is used directly instead.

The feature is only available in the filter panel when the current flow actually
contains rows that can be fingerprinted, so it remains hidden for data that
doesn't benefit from it. Capture ID is also shown as a field in the message
detail dialog, displaying "multiple" when a consolidated parent row is opened.

This submission supersedes the original Homer 7 implementation from PR 
sipcapture/homer-ui#646.

AI has assisted in the development of this feature. I confirm that I have read
and understand the changes and take responsibility for the submission.

### CODE REVIEW

File: src/ui/src/dashboard/flow/FlowItem.tsx [hunk 1]
```diff
@@ -23,73 +42,108 @@ export function FlowItem({
...
+                      if (!setExpandedKey) return
+                      const nextExpanded = showSubItems ? null : itemKey
+                      setExpandedKey(nextExpanded)
+                    }}
+                    aria-label={showSubItems ? 'Collapse consolidated messages' : 'Expand consolidated messages'}
+                    title={showSubItems ? 'Collapse consolidated messages' : 'Expand consolidated messages'}
+                  >
+                    {showSubItems ? '▽' : '▷'}
+                  </span>
+                ) : null}
```
[the remark this answers is not in the document]
Comment (danieldonoghue) [reply src/ui/src/dashboard/flow/FlowItem.tsx]: Fixed in the follow-up commit. Replaced the `<span>` with a `<button type="button">` and added `aria-expanded` to reflect the current open/closed state.

File: src/ui/src/dashboard/flow/flow-data.ts [hunk 2]
```diff
@@ -176,6 +184,131 @@ export function payloadTypeOf(msg: RawMessage): string {
...
+  if (!opts.enabled || items.length === 0) {
+    return items.map((item) => (item.subItems ? { ...item, subItems: undefined } : item))
+  }
+
+  const groups = new Map<string, FlowItemData[]>()
+  const visible: FlowItemData[] = []
+
+  for (const item of items) {
+    const resetItem = item.subItems ? { ...item, subItems: undefined } : item
+    const fp = resetItem.runtimeFingerprint || ''
```
[the remark this answers is not in the document]
Comment (danieldonoghue) [reply src/ui/src/dashboard/flow/flow-data.ts:282]: Fixed. Both branches of `consolidateFlowItems` now always spread `{ ...item, subItems: undefined }` so the original items array is never mutated.

File: src/ui/src/dashboard/flow/flow-data.ts [hunk 3]
```diff
@@ -176,6 +184,131 @@ export function payloadTypeOf(msg: RawMessage): string {
...
+    middle: parent.middle,
+    rightEnd: parent.rightEnd,
+    direction: parent.direction,
+    isRadial: parent.isRadial,
+    isLastHost: parent.isLastHost,
+    arrowStyleSolid: parent.arrowStyleSolid,
+  }
+}
+
+export function consolidateFlowItems(items: FlowItemData[], opts: ConsolidationOptions): FlowItemData[] {
```
[the remark this answers is not in the document]
Comment (danieldonoghue) [reply src/ui/src/dashboard/flow/flow-data.ts:272]: Fixed. Added 15 new tests to `flow-data.test.ts` covering `runtimeFingerprintOf` (empty inputs, consistent hashing, different payloads, pre-existing fingerprint field), `captureIdOf` (top-level field, node_id fallback, data_extra JSON, missing field), and `consolidateFlowItems` (disabled passthrough, consolidation within threshold, threshold exclusion, same capture ID exclusion, different fingerprint exclusion, non-mutation).

File: Dockerfile [hunk 4]
```diff
@@ -1,9 +1,10 @@
...
-FROM golang:bookworm AS builder
+FROM golang:bookworm AS devbase
 
 # Base build deps (no Debian nodejs — bookworm ships Node 18; Vite 7 needs 20.19+ / 22.12+).
-RUN apt-get update && apt-get install -y --no-install-recommends \
+RUN apt-get update && apt-get upgrade -y --no-install-recommends \
+    && apt-get install -y --no-install-recommends \
     git ca-certificates build-essential curl \
     libluajit-5.1-dev \
     && rm -rf /var/lib/apt/lists/*
```
[the remark this answers is not in the document]
Comment (danieldonoghue) [reply Dockerfile]: Fair point, removed. Updated base images are the correct mechanism for addressing OS-level CVEs rather than upgrading packages at build time.

File: src/ui/src/dashboard/CallFlow.tsx [hunk 5]
```diff
@@ -37,6 +39,26 @@ export default function CallFlow({ items, timeZone, onClickMessage }: CallFlowPr
     [filteredItems, timeZone, filters.hostGrouping, locale],
   )
 
+  const canConsolidateCaptureIds = useMemo(
+    () => hasRuntimeFingerprintCandidates(filteredItems),
+    [filteredItems],
+  )
```
[the remark this answers is not in the document]
Comment (danieldonoghue) [reply src/ui/src/dashboard/CallFlow.tsx:45]: Fixed. `canConsolidateCaptureIds` now derives from `flowItems` (which already carries the computed `runtimeFingerprint` field from `buildFlow`) instead of recomputing on `filteredItems`.

File: src/ui/src/dashboard/MessageModal.tsx [hunk 6]
```diff
@@ -147,11 +151,21 @@ function formatDateTime(value, locale, timeZone, dateOnly = false) {
...
+              <dt className="text-[10px] uppercase tracking-wider text-muted-foreground">{META_FIELD_LABELS[field] ?? field}</dt>
+              <dd className="truncate font-mono text-muted-foreground" title={String(overrides[field])}>
+                {String(overrides[field])}
+              </dd>
+            </div>
+          )
+        }
         const raw = data[field]
         const display = raw === undefined
           ? '—'
```
[the remark this answers is not in the document]
Comment (danieldonoghue) [reply src/ui/src/dashboard/MessageModal.tsx:171]: Fixed. Both the override path and the standard rendering path now use `META_FIELD_LABELS[field] ?? field` for the `<dt>` label, so `node_id` consistently displays as "Capture ID" in all cases.

**Comment by danieldonoghue** [comment]
Regarding copilot's callout..
> The PR description says the consolidation feature should be hidden in the filter panel when the current flow has no fingerprint-capable row

the controls are now hidden entirely when no fingerprint-capable rows are present rather than shown disabled, so the saved preference accurately reflects the switch state at all times.

**Comment by danieldonoghue** [comment]
regarding suppressed comments:

- unused CSS: removed
- toolbar "messages" label inaccuracy: The count reflects visible top-level rows with all groups collapsed; this is consistent with how method and payload filters already behave.
- missing tests: added tests covering valid round-trip, invalid boolean value, negative threshold clamping to 0, and non-finite threshold value

**Comment by danieldonoghue** [comment]
All suppressed comments should now have been addressed. note that im using `slice().reverse().find()` instead of `findLast()` to avoid an ES2023 library requirement at this stage.

**Comment by adubovikov** [comment]
## Review follow-up

I reviewed this PR and pushed a small fixup on top of `consolidate-capture-ids` with three tweaks:

1. **Capture ID in message meta** — `MessageModal` now resolves Capture ID via the same `captureIdOf()` path used for consolidation (not only `node_id`). Call Flow also passes `flowItem.captureId` into `metaOverrides` for sub-items; consolidated parents still show `multiple`.
2. **Expand/collapse responsiveness** — `FlowItem` derives `showSubItems` from `expandedKey` directly (no `useEffect` lag) and only renders the toggle when `setExpandedKey` is provided.
3. **Filter toggle gating** — consolidation controls appear only when at least one fingerprint is shared by **two or more distinct** Capture IDs (`canConsolidateFlowItems`), not merely when any fingerprintable row exists.

Branch (fork): https://github.com/adubovikov/homer/tree/review/916-capture-consolidation-fixes  
Commit: https://github.com/adubovikov/homer/commit/9def2dc

Happy to open a PR against this branch, or you can cherry-pick `9def2dc` if useful.

**Review by adubovikov — APPROVED** [review]
LGTM for the Capture-ID consolidation feature.

Follow-up tweaks (Capture ID meta resolution, expand-state sync, stricter toggle gating) are ready in a separate PR after this merges.
